qa/common/common.sh

```diff
@@ -2,50 +2,71 @@
-
-# determine hostname of Salt Master
-MASTER_MINION_SLS=/srv/pillar/ceph/master_minion.sls
-if test -s $MASTER_MINION_SLS ; then
-    SALT_MASTER=$(cat $MASTER_MINION_SLS | \
-                 sed 's/.*master_minion:[[:blank:]]*\(\w\+\)[[:blank:]]*/\1/' | \
-                 grep -v '^$')
-else
-    echo "Could not determine the Salt Master from DeepSea pillar data. Is DeepSea installed?"
-    exit 1
-fi
-
-# show which repos are active/enabled
-zypper lr -upEP
-
-# show salt RPM version in log and fail if salt is not installed
-rpm -q salt-master
-rpm -q salt-minion
-rpm -q salt-api
-
-# show deepsea RPM version in case deepsea was installed from RPM
-rpm -q deepsea || true
-
-# set deepsea_minions to * - see https://github.com/SUSE/DeepSea/pull/526
-# (otherwise we would have to set deepsea grain on all minions)
-echo "deepsea_minions: '*'" > /srv/pillar/ceph/deepsea_minions.sls
-cat /srv/pillar/ceph/deepsea_minions.sls
-
-# get list of minions
-if type salt-key > /dev/null 2>&1; then
-    MINIONS_LIST=$(salt-key -L -l acc | grep -v '^Accepted Keys')
-else
-    echo "Cannot find salt-key. Is Salt installed? Is this running on the Salt Master?"
-    exit 1
-fi
 
+function global_test_init {
+    #
+    # determine hostname of Salt Master
+    SALT_MASTER=$(hostname)
+    #
+    # show which repos are active/enabled
+    zypper lr -upEP
+    #
+    # show salt RPM version in log and fail if salt is not installed
+    rpm -q salt-master
+    rpm -q salt-minion
+    rpm -q salt-api
+    #
+    # show deepsea RPM version in case deepsea was installed from RPM
+    rpm -q deepsea || true
+    #
+    # set deepsea_minions to * - see https://github.com/SUSE/DeepSea/pull/526
+    # (otherwise we would have to set deepsea grain on all minions)
+    echo "deepsea_minions: '*'" > /srv/pillar/ceph/deepsea_minions.sls
+    cat /srv/pillar/ceph/deepsea_minions.sls
+    #
+    # get list of minions
+    if type salt-key > /dev/null 2>&1; then
+        MINIONS_LIST=$(salt-key -L -l acc | grep -v '^Accepted Keys')
+    else
+        echo "Cannot find salt-key. Is Salt installed? Is this running on the Salt Master?"
+        exit 1
+    fi
+}
+
+function ping_minions_until_all_respond {
+    local NUM_MINIONS="$1"
+    local RESPONDING=""
+    for i in {1..20} ; do
+        RESPONDING=$(salt '*' test.ping 2>/dev/null | grep True 2>/dev/null | wc --lines)
+        echo "Of $NUM_MINIONS total minions, $RESPONDING are responding"
+        test "$NUM_MINIONS" -eq "$RESPONDING" && break
+        sleep 10
+    done
+}
+
+function update_salt {
+    # make sure we are running the latest Salt before Stage 0 starts,
+    # otherwise Stage 0 will update Salt and then fail with cryptic
+    # error messages
+    TOTAL_NODES=$(json_total_nodes)
+    salt '*' cmd.run 'zypper -n in -f python3-salt salt salt-api salt-master salt-minion'
+    systemctl restart salt-api.service
+    systemctl restart salt-master.service
```

Comment:
After restarting the master, there's the chance that some minions take longer to re-connect to the master which means that the succeeding call will not reach all the minions.

To be extra sure me might add an extra `sleep` here?

Reply:
Yeah, I encountered that in some runs. That's why I have the `ping_minions_until_all_respond` call three lines down. But I'll move the sleep up so it's between master restart and minion restart.